lib/install/engine/cli/cli.go

```diff
@@ -0,0 +1,283 @@
+			System: storage.SystemVariables{
+				Docker: r.config.Docker,
+			},
+			OnPrem: storage.OnPremVariables{
+				PodCIDR:     r.config.PodCIDR,
+				ServiceCIDR: r.config.ServiceCIDR,
+				VxlanPort:   r.config.VxlanPort,
+			},
+		},
+		Profiles: install.ServerRequirements(*r.config.Flavor),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	operation, err := r.Operator.GetSiteOperation(*key)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return operation, nil
+}
+
+func (r *executor) waitForAgents(operation ops.SiteOperation) error {
+	ctx, cancel := context.WithTimeout(r.ctx, 5*time.Minute)
+	defer cancel()
+	b := utils.NewUnlimitedExponentialBackOff()
+	b.MaxInterval = 5 * time.Second
+	var oldReport *ops.AgentReport
+	err := utils.RetryWithInterval(ctx, b, func() error {
+		report, err := r.Operator.GetSiteInstallOperationAgentReport(operation.Key())
+		if err != nil {
+			return trace.Wrap(err, "failed to get agent report")
+		}
+		old := oldReport
+		oldReport = report
+		if !r.canContinue(old, report) {
+			return trace.BadParameter("cannot continue")
+		}
+		r.WithField("report", report).Info("Installation can proceed.")
+		err = libinstall.UpdateOperationState(r.Operator, operation, *report)
+		return trace.Wrap(err)
+	})
+	return trace.Wrap(err)
+}
+
+// canContinue returns true if all agents have joined and the installation can start
+func (r *executor) canContinue(old, new *ops.AgentReport) bool {
+	// See if any new nodes have joined or left since previous agent report.
+	joined, left := new.Diff(old)
+	for _, server := range joined {
+		r.PrintStep(color.GreenString("Successfully added %q node on %v",
+			server.Role, utils.ExtractHost(server.AdvertiseAddr)))
+	}
+	for _, server := range left {
+		r.PrintStep(color.YellowString("Node %q on %v has left",
+			server.Role, utils.ExtractHost(server.AdvertiseAddr)))
+	}
+	// See if the current agent report satisfies the selected flavor.
+	needed, extra := new.MatchFlavor(r.config.Flavor)
+	if len(needed) == 0 && len(extra) == 0 {
+		r.PrintStep(color.GreenString("All agents have connected!"))
+		return true
+	}
+	// If there were no changes compared to previous report, do not
+	// output anything.
+	if len(joined) == 0 && len(left) == 0 {
+		return false
+	}
+	// Dump the table with remaining nodes that need to join.
+	r.PrintStep("Please execute the following join commands on target nodes:\n%v",
+		formatProfiles(needed, r.config.AdvertiseAddr, r.config.Token.Token))
+	// If there are any extra agents with roles we don't expect for
+	// the selected flavor, they need to leave.
+	for _, server := range extra {
+		r.PrintStep(color.RedString("Node %q on %v is not a part of the flavor, shut it down",
+			server.Role, utils.ExtractHost(server.AdvertiseAddr)))
+	}
+	return false
+}
+
+// Engine implements command line-driven installation workflow
+type Engine struct {
+	// Config specifies the engine's configuration
+	Config
+}
+
+type executor struct {
```

Comment:
I know it looks weird but I tried to avoid setting attributes as side-effects of some other operation and since the engine only gets hands on some of the data in Execute, it creates and uses an `executor` to drive the operation.